Source/Global/perform_env.cpp

```diff
@@ -210,32 +208,130 @@ Result<HC_UNIQUE_PTR<HC_PERFORM_ENV>> HC_PERFORM_ENV::Initialize(HCInitArgs* arg
 #elif HC_PLATFORM == HC_PLATFORM_ANDROID
     auto initAndroidResult = AndroidPlatformContext::Initialize(args);
     RETURN_IF_FAILED(initAndroidResult.hr);
-
-    performEnv.reset(new (a.allocate(1)) HC_PERFORM_ENV);
     performEnv->androidPlatformContext = initAndroidResult.ExtractPayload();
 
 #else
-    UNREFERENCED_PARAMETER(a);
     RETURN_HR_IF(E_INVALIDARG, args);
 #endif
 
     return std::move(performEnv);
 }
 
+struct HC_PERFORM_ENV::HttpPerformContext
+{
+    HttpPerformContext(HC_PERFORM_ENV* _env, HCCallHandle _callHandle, XAsyncBlock* _clientAsyncBlock) :
+        env{ _env },
+        callHandle{ _callHandle },
+        clientAsyncBlock{ _clientAsyncBlock },
+        internalAsyncBlock{ nullptr, this, HC_PERFORM_ENV::HttpPerformComplete }
+    {
+    }
+
+    ~HttpPerformContext()
+    {
+        XTaskQueueCloseHandle(internalAsyncBlock.queue);
+    }
+
+    HC_PERFORM_ENV* env{};
+    HCCallHandle const callHandle{};
+    XAsyncBlock* const clientAsyncBlock{};
+    XAsyncBlock internalAsyncBlock{};
+};
+
+HRESULT HC_PERFORM_ENV::HttpCallPerformAsyncShim(HCCallHandle call, XAsyncBlock* async)
+{
+    auto performContext = http_allocate_shared<HttpPerformContext>(this, call, async);
+    {
+        std::unique_lock<std::mutex> lock{ m_mutex };
+        m_activeHttpRequests[performContext->callHandle] = performContext;
+    }
+    return XAsyncBegin(async, performContext.get(), nullptr, __FUNCTION__, HttpPerformAsyncShimProvider);
+}
+
 HRESULT HC_PERFORM_ENV::CleanupAsync(HC_UNIQUE_PTR<HC_PERFORM_ENV>&& env, XAsyncBlock* async) noexcept
 {
     RETURN_IF_FAILED(XAsyncBegin(async, env.get(), __FUNCTION__, __FUNCTION__, CleanupAsyncProvider));
     env.release();
     return S_OK;
 }
 
+HRESULT CALLBACK HC_PERFORM_ENV::HttpPerformAsyncShimProvider(XAsyncOp op, const XAsyncProviderData* data)
+{
+    HttpPerformContext* performContext{ static_cast<HttpPerformContext*>(data->context) };
+    HC_PERFORM_ENV* env{ performContext->env };
+
+    switch (op)
+    {
+    case XAsyncOp::Begin:
+    {
+        XTaskQueuePortHandle workPort{};
+        RETURN_IF_FAILED(XTaskQueueGetPort(data->async->queue, XTaskQueuePort::Work, &workPort));
```

Comment:
>data->async->queue [](http://example.com/codeflow?start=43&length=18)

if you call in with a null queue, does this work?

Reply:
no, but queue can't be null at this point.  Even if client's XAsyncBlock had a null queue, XAsync will pass the process queue to the provider

Reply:
you may want to assert to document the expectation

Reply:
added assert